Maelstrom/AbilityShuffle.cs

```diff
@@ -23,82 +23,123 @@ public static List<JunctionableGF> Randomise(FileSource mainSource, int seed, St
+        {
+            var includedAbilities = Abilities.Where(a => !a.ItemExclusive || settings.GfAbilitiesIncludeItemOnly).ToList();
+            var abilityLimit = 21;
+
+            try
+            {
+                abilityLimit = Int32.Parse(settings.GfAbilitiesLimit);
+            } catch (Exception)
+            { }
+
+            for (int gfId = 0; gfId < 16; gfId++)
+            {
+                // clear auto-unlocked abilities
+                init.GFs[gfId].Abilities = new BitArray(init.GFs[gfId].Abilities.Length, false);
+
+                var unusedAbilities = includedAbilities.Select(a => a.AbilityID).ToList();
+                for (int learnSlotIndex = 0; learnSlotIndex < abilityLimit; learnSlotIndex++)
                 {
-                    // clear auto-unlocked abilities
-                    init.GFs[i].Abilities = new BitArray(init.GFs[i].Abilities.Length, false);
+                    if (learnSlotIndex < 4 && settings.GfAbilitiesBasics)
+                    {
+                        AddBasicAbilities(kernel.JunctionableGFs[gfId].Abilities, init, gfId, unusedAbilities);
 
-                    var unusedAbilities = includedAbilities.Select(a => a.AbilityID).ToList();
-                    for (int j = 0; j < 21; j++)
+                        learnSlotIndex = 3;
+                    }
+                    else
                     {
-                        // guarantee basic commands (magic, gf, draw, item)
-                        if (j < 4 && settings.GfAbilitiesBasics)
-                        {
-                            var id = j + 20;
-                            kernel.JunctionableGFs[i].Abilities[j].Ability = (byte)id;
-                            kernel.JunctionableGFs[i].Abilities[j].Unlocker = 0;
-                            init.GFs[i].Abilities[id] = true;
-                            unusedAbilities.Remove(id);
-                            continue;
-                        }
-
-                        // roll abilities
-                        var ability = unusedAbilities[random.Next(unusedAbilities.Count)];
-                        kernel.JunctionableGFs[i].Abilities[j].Ability = (byte)ability;
-                        kernel.JunctionableGFs[i].Abilities[j].Unlocker = 0;
-                        init.GFs[i].Abilities[ability] = (ability >= 20 && ability <= 23);
-                        unusedAbilities.Remove(ability);
+                        AddRandomAbility(random, kernel.JunctionableGFs[gfId].Abilities, learnSlotIndex, init.GFs[gfId], unusedAbilities);
                     }
+                }
 
-                    // sort abilities
-                    kernel.JunctionableGFs[i].Abilities = kernel.JunctionableGFs[i].Abilities.OrderBy(a => a.Ability == 0 ? byte.MaxValue : a.Ability).ToArray();
-
-                    // clear ability being learned
-                    init.GFs[i].CurrentAbility = 0;
+                //This duplicates the first ability for the remaining slots. It's hacky, but it works.
```

Comment:
It's odd - when you're learning the ability, you can pretty clearly see a 0 behind the actual AP you've earned toward the ability, indicating that there are multiple copies of it. After learning the ability, there doesn't seem to be an issue. I will try amnesia greens on duplicated abilities (learned and unlearned).

Reply:
So yeah, this is pretty cursed. Using Amnesia Greens on a learned ability appears to unlearn the ability and reset it to 0 AP. For an unlearned ability, this appears to do nothing. In the configuration I tested, there are nine copies of the same ability. No amount of Amnesia Greens would remove the duplicated ability, so this definitely doesn't work.

An alternative would be figuring out how the unlocking/hidden abilities mechanic works and hide a bunch of abilities behind an ability we don't have. Maybe the Empty ability and remove that from the pool of random abilities? I think unlockers have to be chained, so there's plenty of opportunity for someone to learn something from an item and suddenly have a random ability available to learn. Also I have basically zero idea how that works, so this would require a lot of experimentation.

I'd recommend either shelving this for now, or making a note in the UI of the issues that it has. Let me know which you prefer and I'll sort it out.